connman/plugins/vpn.c

```diff
@@ -1590,57 +1590,88 @@ static void set_route(struct connection_data *data, struct vpn_route *route)
 static int save_route(GHashTable *routes, int family, const char *network,
 			const char *netmask, const char *gateway);
 
-static int add_network_route(struct connection_data *data)
+
+static int set_network_route(struct connection_data *data, struct vpn_route *rt)
 {
-	struct vpn_route rt = { 0, };
 	int err;
 
-	if (!data)
+	if (!data || !rt)
 		return -EINVAL;
 
-	rt.family = connman_provider_get_family(data->provider);
-	switch (rt.family) {
-	case PF_INET:
-		err = connman_inet_get_route_addresses(data->index,
-					&rt.network, &rt.netmask, &rt.gateway);
-		break;
-	case PF_INET6:
-		err = connman_inet_ipv6_get_route_addresses(data->index,
-					&rt.network, &rt.netmask, &rt.gateway);
-		break;
-	default:
-		connman_error("invalid protocol family %d", rt.family);
-		return -EINVAL;
-	}
-
 	DBG("network %s gateway %s netmask %s for provider %p",
-						rt.network, rt.gateway, rt.netmask,
-						data->provider);
-
-	if (err) {
-		connman_error("cannot get network/gateway/netmask for %p",
-							data->provider);
-		goto out;
-	}
+					rt->network, rt->gateway, rt->netmask,
+					data->provider);
 
-	err = save_route(data->server_routes, rt.family, rt.network, rt.netmask,
-				rt.gateway);
+	err = save_route(data->server_routes, rt->family, rt->network,
+						rt->netmask, rt->gateway);
 	if (err) {
 		connman_warn("failed to add network route for provider"
 					"%p", data->provider);
-		goto out;
+		return err;
 	}
 
-	set_route(data, &rt);
-
-out:
-	g_free(rt.network);
-	g_free(rt.netmask);
-	g_free(rt.gateway);
+	set_route(data, rt);
 
 	return 0;
 }
 
+static void free_network_route(struct vpn_route *rt)
+{
+	g_free(rt->network);
+	rt->network = NULL;
+
+	g_free(rt->netmask);
+	rt->network = NULL;
+
+	g_free(rt->gateway);
+	rt->network = NULL;
+}
+
+static int add_network_route(struct connection_data *data)
+{
+	struct vpn_route rt = { 0, };
+	int err = 0;
+
+	if (!data)
+		return -EINVAL;
+
+	if (connman_provider_get_family(data->provider, AF_INET)) {
+		rt.family = AF_INET;
+		err = connman_inet_get_route_addresses(data->index,
+					&rt.network, &rt.netmask,
+					&rt.gateway);
+		if (err)
+			connman_error("cannot get IPv4 network/gateway/netmask "
+						"for %p/%s", data->provider,
+						data->ident);
+		else
+			err = set_network_route(data, &rt);
+
+		free_network_route(&rt);
+	}
+
+	if (connman_provider_get_family(data->provider, AF_INET6)) {
+		rt.family = AF_INET6;
+		err = connman_inet_ipv6_get_route_addresses(data->index,
```

Comment:
Success here can hide error from earlier op. Does it matter?

Reply:
Not really sure if it matters, if either of the routes gets set, the end result is half-correct if the VPN had both IP family routes to set.  Combining errors in these cases is not commonly used, there are some cases where either is ok = ok. Maybe just to report the error to logs instead of returning an error would be better approach if either fails?

Reply:
08ede161a7003bcf4cce4b619ea3ed15be1df023 report better